waitFor(QCoro::Task<T>): Do not require T to be default constructible

If you use `QCoro::waitFor` with a `QCoro::Task<T>`, the type `T` must be default constructible. It would nice to be able to use this function with types that do not have a default constructor.

`T` is required to be default constructible because in `impl/waitfor.h:76` a default `T` is constructed, whose value is immediately overwritten in the call to `QCoro::details::runCoroutine()`.

Example compile error from gcc with `T` being a boost::outcome::result<int, std::string>`:

```
In file included from ../../../deps/include/qcoro5/qcoro/qcorotask.h:568,
                 from ../../../deps/include/qcoro5/QCoro/QCoroTask:1,
                 from ../../../test.h:15,
../../../deps/include/qcoro5/qcoro/impl/waitfor.h: In instantiation of 'T QCoro::detail::waitFor(Awaitable&&) [with T = boost::outcome_v2::basic_result<int>, std::__cxx11::basic_string<char>, boost::outcome_v2::policy::exception_ptr_rethrow<int>, std::__cxx11::basic_string<char>, void> >; Awaitable = QCoro::Task<boost::outcome_v2::basic_result<int>, std::__cxx11::basic_string<char>, boost::outcome_v2::policy::exception_ptr_rethrow<int>, std::__cxx11::basic_string<char>, void> > >]':
../../../deps/include/qcoro5/qcoro/impl/waitfor.h:97:30:   required from 'T QCoro::waitFor(QCoro::Task<T>&&) [with T = boost::outcome_v2::basic_result<int>, std::__cxx11::basic_string<char>, boost::outcome_v2::policy::exception_ptr_rethrow<int>, std::__cxx11::basic_string<char>, void> >]'
RemoteCollectorDatabase.h:27:30:   required from here
../../../deps/include/qcoro5/qcoro/impl/waitfor.h:76:11: error: use of deleted function 'boost::outcome_v2::basic_result<R, S, NoValuePolicy>::basic_result() [with R = int>; S = std::__cxx11::basic_string<char>; NoValuePolicy = boost::outcome_v2::policy::exception_ptr_rethrow<int>, std::__cxx11::basic_string<char>, void>]'
   76 |         T result;
      |           ^~~~~~
In file included from ../../../deps/include/boost/outcome/std_result.hpp:34,
                 from ../../../deps/include/boost/outcome/boost_result.hpp:60,
                 from ../../../deps/include/boost/outcome/result.hpp:34,
                 from ../../../libsceptre/test.h:14,
../../../deps/include/boost/outcome/basic_result.hpp:372:3: note: declared here
  372 |   basic_result() = delete;
```

--- a/qcoro/impl/waitfor.h
+++ b/qcoro/impl/waitfor.h
@@ -12,6 +12,8 @@
 
 #include <QEventLoop>
 
+#include <optional>
+
 namespace QCoro
 {
 
@@ -51,9 +53,9 @@ Task<> runCoroutine(WaitContext &context, Awaitable &&awaitable) {
 }
 
 template<typename T, Awaitable Awaitable>
-Task<> runCoroutine(WaitContext &context, T &result, Awaitable &&awaitable) {
+Task<> runCoroutine(WaitContext &context, std::optional<T> &result, Awaitable &&awaitable) {
     try {
-        result = co_await awaitable;
+        result.emplace(co_await awaitable);
     } catch (...) {
         context.exception = std::current_exception();
     }
@@ -73,7 +75,7 @@ T waitFor(Awaitable &&awaitable) {
             std::rethrow_exception(context.exception);
         }
     } else {
-        T result;
+        std::optional<T> result;
         runCoroutine(context, result, std::forward<Awaitable>(awaitable));
         if (!context.coroutineFinished) {
             context.loop.exec();
@@ -81,7 +83,7 @@ T waitFor(Awaitable &&awaitable) {
         if (context.exception) {
             std::rethrow_exception(context.exception);
         }
-        return result;
+        return *result;
     }
 }
 

--- a/qcoro/impl/waitfor.h
+++ b/qcoro/impl/waitfor.h
@@ -53,7 +53,18 @@ Task<> runCoroutine(WaitContext &context, Awaitable &&awaitable) {
 }
 
 template<typename T, Awaitable Awaitable>
-Task<> runCoroutine(WaitContext &context, std::optional<T> &result, Awaitable &&awaitable) {
+Task<> runCoroutine(WaitContext &context, T &result, Awaitable &&awaitable) {
+    try {
+        result = co_await awaitable;
+    } catch (...) {
+        context.exception = std::current_exception();
+    }
+    context.coroutineFinished = true;
+    context.loop.quit();
+}
+
+template<typename T, Awaitable Awaitable>
+Task<> runCoroutineOpt(WaitContext &context, std::optional<T> &result, Awaitable &&awaitable) {
     try {
         result.emplace(co_await awaitable);
     } catch (...) {
@@ -75,15 +86,27 @@ T waitFor(Awaitable &&awaitable) {
             std::rethrow_exception(context.exception);
         }
     } else {
-        std::optional<T> result;
-        runCoroutine(context, result, std::forward<Awaitable>(awaitable));
-        if (!context.coroutineFinished) {
-            context.loop.exec();
+        if constexpr (std::is_default_constructible_v<T>) {
+            T result;
+            runCoroutine(context, result, std::forward<Awaitable>(awaitable));
+            if (!context.coroutineFinished) {
+                context.loop.exec();
+            }
+            if (context.exception) {
+                std::rethrow_exception(context.exception);
+            }
+            return result;
+        } else {
+            std::optional<T> result;
+            runCoroutineOpt(context, result, std::forward<Awaitable>(awaitable));
+            if (!context.coroutineFinished) {
+                context.loop.exec();
+            }
+            if (context.exception) {
+                std::rethrow_exception(context.exception);
+            }
+            return *result;
         }
-        if (context.exception) {
-            std::rethrow_exception(context.exception);
-        }
-        return *result;
     }
 }
 

--- a/qcoro/impl/waitfor.h
+++ b/qcoro/impl/waitfor.h
@@ -83,7 +83,7 @@ T waitFor(Awaitable &&awaitable) {
         if (context.exception) {
             std::rethrow_exception(context.exception);
         }
-        return *result;
+        return std::move(*result);
     }
 }
 

--- a/qcoro/impl/waitfor.h
+++ b/qcoro/impl/waitfor.h
@@ -53,18 +53,7 @@ Task<> runCoroutine(WaitContext &context, Awaitable &&awaitable) {
 }
 
 template<typename T, Awaitable Awaitable>
-Task<> runCoroutine(WaitContext &context, T &result, Awaitable &&awaitable) {
-    try {
-        result = co_await awaitable;
-    } catch (...) {
-        context.exception = std::current_exception();
-    }
-    context.coroutineFinished = true;
-    context.loop.quit();
-}
-
-template<typename T, Awaitable Awaitable>
-Task<> runCoroutineOpt(WaitContext &context, std::optional<T> &result, Awaitable &&awaitable) {
+Task<> runCoroutine(WaitContext &context, std::optional<T> &result, Awaitable &&awaitable) {
     try {
         result.emplace(co_await awaitable);
     } catch (...) {
@@ -86,27 +75,15 @@ T waitFor(Awaitable &&awaitable) {
             std::rethrow_exception(context.exception);
         }
     } else {
-        if constexpr (std::is_default_constructible_v<T>) {
-            T result;
-            runCoroutine(context, result, std::forward<Awaitable>(awaitable));
-            if (!context.coroutineFinished) {
-                context.loop.exec();
-            }
-            if (context.exception) {
-                std::rethrow_exception(context.exception);
-            }
-            return result;
-        } else {
-            std::optional<T> result;
-            runCoroutineOpt(context, result, std::forward<Awaitable>(awaitable));
-            if (!context.coroutineFinished) {
-                context.loop.exec();
-            }
-            if (context.exception) {
-                std::rethrow_exception(context.exception);
-            }
-            return *result;
+        std::optional<T> result;
+        runCoroutine(context, result, std::forward<Awaitable>(awaitable));
+        if (!context.coroutineFinished) {
+            context.loop.exec();
         }
+        if (context.exception) {
+            std::rethrow_exception(context.exception);
+        }
+        return *result;
     }
 }
 

--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -114,6 +114,7 @@ qcoro_add_test(testconstraints)
 qcoro_add_test(qfuture LINK_LIBRARIES Qt${QT_VERSION_MAJOR}::Concurrent)
 qcoro_add_test(qcorogenerator)
 qcoro_add_test(qcoroasyncgenerator)
+qcoro_add_test(qcorowaitfor)
 
 if (QCORO_WITH_QTDBUS)
     qcoro_add_dbus_test(qdbuspendingcall)

--- a/tests/qcorowaitfor.cpp
+++ b/tests/qcorowaitfor.cpp
@@ -1,0 +1,62 @@
+// SPDX-FileCopyrightText: 2024 Joey Richey <joey@dogsplayingpoker.org>
+//
+// SPDX-License-Identifier: MIT
+
+#include "testobject.h"
+
+#include "qcorotimer.h"
+
+#include <chrono>
+
+class QCoroWaitForTest : public QCoro::TestObject<QCoroWaitForTest> {
+    Q_OBJECT
+
+private:
+    QCoro::Task<void> testPrimitiveType_coro(QCoro::TestContext ctx)
+    {
+        ctx.setShouldNotSuspend();
+        auto task_test = []() -> QCoro::Task<int> {
+            co_return 7;
+        };
+
+        const int ret = QCoro::waitFor(task_test());
+        QCORO_VERIFY(ret == 7);
+    }
+
+    QCoro::Task<void> testDefaultConstructible_coro(QCoro::TestContext ctx)
+    {
+        ctx.setShouldNotSuspend();
+        auto task_test = []() -> QCoro::Task<std::string> {
+            co_return "seven";
+        };
+
+        const std::string ret = QCoro::waitFor(task_test());
+        QCORO_VERIFY(ret == "seven");
+    }
+
+    QCoro::Task<void> testNonDefaultConstructible_coro(QCoro::TestContext ctx)
+    {
+        ctx.setShouldNotSuspend();
+
+        struct test_struct {
+            explicit test_struct(int i_) : i(i_) {}
+            int i;
+        };
+
+        auto task_test = []() -> QCoro::Task<test_struct> {
+            co_return test_struct(7);
+        };
+
+        const test_struct ret = QCoro::waitFor(task_test());
+        QCORO_VERIFY(ret.i == 7);
+    }
+
+private Q_SLOTS:
+    addTest(PrimitiveType)
+    addTest(DefaultConstructible)
+    addTest(NonDefaultConstructible)
+};
+
+QTEST_GUILESS_MAIN(QCoroWaitForTest)
+
+#include "qcorowaitfor.moc"

